sql: fix TestRandomSyntaxSelect

ZhouXing19 opened the pull request:
Ignore the error if the given random query needs comparison between tuple and
string.

Fixes https://github.com/cockroachdb/cockroach/issues/72257

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73263)
<!-- Reviewable:end -->

rafiss commented:
Hm, I think we should be able to fix this and handle the error instead of continuing to return this unhandled stack trace.

In this case it's new functionality on the `master` branch, so this indicates that we forgot something when adding this feature. Feel free to re-assign to me if you are not sure where to look.

```
root@:26257/defaultdb>         SELECT ROW ( ) IS NAN ident FROM ident ;
ERROR: internal error: comparison overload not found (eq, tuple, string)
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/logical_props_builder.go:1640: BuildSharedProps()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/logical_props_builder.go:1546: buildProjectionsItemProps()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:8588: PopulateProps()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:12257: ConstructProjectionsItem()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:55: constructProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:36: constructProjectForScope()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1029: buildSelectClause()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:930: buildSelectStmtWithoutParens()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:903: func1()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/with.go:115: processWiths()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:902: buildSelect()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:281: buildStmt()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:241: buildStmtAtRoot()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:215: Build()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:521: buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:194: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1264: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:962: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:700: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1752: func1()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1754: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:683: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:666: func1()
runtime/asm_amd64.s:1581: goexit()
```

ZhouXing19 committed:
sql: fix TestRandomSyntaxSelect

Ignore the error if the given random query needs comparison between tuple and
string.

Fixes #72257

Release note: None

ZhouXing19 commented:
@rafiss Do you mean that we should support the `ROW ( ) IS NAN` syntax, or we should just handle the error (by adding it to some ignore list)?

rafiss commented:
I mean we should return a nicer error message if the query is executed:

For example, if you run this query, notice how the error message is formatted nicely and doesn't look scary.
```
root@:26257/defaultdb> select 'cat' * 'dog';
ERROR: unsupported binary operator: <string> * <string>
SQLSTATE: 22023
```

However, a message like `ERROR: internal error: comparison overload not found (eq, tuple, string)` means that our code ran into an unexpected case. A user who sees this (and the long stack trace) might be worried that something more serious was broken. Whenever we show a stack trace with an error, it means that something unexpected happened. It's always better to handle something we don't support as an expected error, so the behavior is well-defined.

otan commented:
given rafi's PR i'll close this one!
